Ensure there are always mass matrix contributors for explicit solves

lindsayad opened the pull request:
Closes https://github.com/idaholab/moose/issues/29834

lindsayad committed:
Check for missing mass matrix contributors

This turned into a much larger project with the addition of `ResidualObject`
to `MooseObjectWarehouse` which necessitated removing `FEProblemBase` and
`NonlinearSystemBase` headers from files in order to avoid circular includes

Refs #28053

lindsayad committed:
Revert "Changed pc type from hypre to none for failing tests. Ref #28053."

This reverts commit de949efa22d6aeac0da4665b2d92889e3bf8bf1e.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2677334/), step Docs: sync website on 22ce188 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29833/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2681618/), step Clang format on 43d9ca9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29833/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29833/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5fce0ac136837e85742ff9c1539c66314ba1f773`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2683659/), step Clang format on ec9143b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/29833/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/29833/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2de6bfcb393c5dae820219bc9723f5d572a60b96`

lindsayad committed:
Add .cache to .gitignore

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2685506/), step Generate coverage on 5f8173a wanted to post the following:


## Framework coverage

Inconsistent report tags were found between the head and base reports.
This can happen when reports are missing from either the head or the base.

Inconsistent tags:
  framework-torch-gpu
[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29833/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>c8ca62</th>
      <th colspan=4>#29833 5f8173</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.83%</td>
      <td>84.85%</td>
      <td>+0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>17777</td>
      <td>17780</td>
      <td>+3</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3178</td>
      <td>3175</td>
      <td>-3</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29833/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29833/coverage/navier_stokes)

### Solid mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>c8ca62</th>
      <th colspan=4>#29833 5f8173</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.98%</td>
      <td>84.98%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>28631</td>
      <td>28633</td>
      <td>+2</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>5059</td>
      <td>5059</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29833/coverage_diff/solid_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29833/coverage/solid_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29833/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on framework/src/timeintegrators/ExplicitTimeIntegrator.C:
```diff
@@ -63,6 +69,76 @@ void
+          if (_nl->getNodalKernelWarehouse()
+                  .getMatrixTagObjectWarehouse(mass_matrix_tag_id, 0)
+                  .hasVariableObjects(var_id))
+            continue;
```
HDG Kernels?

lindsayad replied:
these inherit from `KernelBase` and are contained within the kernel warehouse

GiudGiud reviewed (commented):
great cleanup

lindsayad commented on framework/src/timeintegrators/ExplicitTimeIntegrator.C:
```diff
@@ -63,6 +69,76 @@ void
+
+    const auto mass_matrix_tag_id = massMatrixTagID();
+    std::set<TagID> matrix_tags = {mass_matrix_tag_id};
+    auto fv_object_starting_query = _fe_problem.theWarehouse()
```
because it's only used for querying FV objects. FE objects are all from `MooseObjectWarehouses`, not `theWarehouse`

lindsayad committed:
Apply suggestions from code review

- Fix typos
- Add comment regarding the use of `ResidualObject` in `MooseObjectWarehouse`

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/2688699/) on e439337 : invalidated by @lindsayad

moosebuild commented:
Job [Documentation controlled apps](https://civet.inl.gov/job/2688701/) on e439337 : invalidated by @lindsayad

moosebuild commented:
Job [Internal app tests](https://civet.inl.gov/job/2688704/) on e439337 : invalidated by @lindsayad

moosebuild commented:
Job [Documentation internal apps](https://civet.inl.gov/job/2688702/) on e439337 : invalidated by @lindsayad
